### PR TITLE
storm: make Apollo Cache persistor use preloaded data

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

Addresses [the issue](https://github.com/sourcegraph/sourcegraph/pull/48463#issuecomment-1454894102) spotted by @vovakulikov in https://github.com/sourcegraph/sourcegraph/pull/48463. 

- The Apollo Client cache persistor is updated to use the preloaded data on reads instead of only data available in `localStorage`. 
- The `temporarySettings` query is no longer persisted to the `localStorage`.

## Test plan

Use repro instructions shared by @vovakulikov:

1. Read a temporary settings field,
2. See one value of this field (let's say we have a boolean flag, and its value is `true`).
3. Change this flag value from a different browser or via the API console (editTemporalSettings mutation).
4. Refresh the page and see that `window.context.temporarySettings` has a correct new value.
5. But `useTemporarySettings` returns the old value because it's using a cached value from the local storage that persistMapper provided for the Apollo cache.

## App preview:

- [Web](https://sg-web-vb-storm-5.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.01% (+0.22 kb) | 0.00% (+0.22 kb) | 0.00% (0.00 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/vb/storm-5/compare-report.html;tab=live_object) for a full comparison between the commits c42e284 and 4440938 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Review by sashaostrikov — APPROVED** [review]
LGTM!

File: client/web/src/backend/getPersistentCache.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * to avoid keeping private information in localStorage after logout.
+ */
+const getApolloPersistCacheKey = (isAuthenticated: boolean): string =>
+    `apollo-cache-persist-${isAuthenticated ? 'authenticated' : 'anonymous'}`
+
+/**
+ * Persistence storage is based on `localStorage`, which uses the
+ * data preloaded on the server and transferred to the client via `window.context`.
+ */
+class LocalStorageWrapper implements PersistentStorage<CacheObject | null> {
```
Comment (vovakulikov) [inline client/web/src/backend/getPersistentCache.ts:19]: I'm not sure about this wrapper, I see that in the getItem method, we explicitly enforce using Temporary Settings from window.context object, but it looks like we can just fix persistMapper itself in order to getting from the local storage only queries that we want to persist (ignoring context-based temporally settings that we already have in the cache) 

I may missing something here, let me know if it's not possible

File: client/web/src/backend/getPersistentCache.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    public getItem(key: string): CacheObject | null {
+        const maybeData = window.localStorage.getItem(key) ?? '{}'
+        const persistedData = JSON.parse(maybeData) as CacheObject
+
+        const hydratedData: CacheObject = {
+            ...persistedData,
+            [ROOT_QUERY_KEY]: {
+                __typename: 'Query',
+                ...persistedData[ROOT_QUERY_KEY],
+                temporarySettings: window.context.temporarySettings,
```
Comment (vovakulikov) [inline client/web/src/backend/getPersistentCache.ts:37]: Minor: If we have to have this wrapper, I think we could pass context temporal settings as a parameter to the constructor (as we do for the same level function getApolloPersistCacheKey with isAuthenticated info)

Comment (valerybugakov) [reply client/web/src/backend/getPersistentCache.ts]: I like that, updated!

File: client/web/src/backend/getPersistentCache.ts (hunk 1 above)
Comment (valerybugakov) [reply client/web/src/backend/getPersistentCache.ts:19]: Discussed on the Zoom call. We didn't find a more idiomatic way to provide preloaded data.

**Pull request merged**